bin/i18n/curriculum_sync_utils.rb

```diff
@@ -0,0 +1,204 @@
+  #       "name": "...",
+  #     },
+  #     ...
+  #   },
+  #   "activity_sections": {
+  #     "ef4f13d4-c5b0-4778-8ca1-359132023d82": {
+  #       "progression_name": "..."
+  #     },
+  #     ...
+  #   },
+  #   "resources": {
+  #     "url": {
+  #       "name": "..."
+  #     },
+  #     ...
+  #   }
+  # }
+  def self.flatten(object_hash, serializer, name)
+    results = {}
+    object_hash.each do |key, object|
+      object = object.symbolize_keys
+
+      # store attributes directly on the results for this object
+      attributes = object.slice(*serializer._attributes)
+      if attributes.present?
+        results[name] ||= {}
+        results[name][key] = attributes
+      end
+
+      # recursively process "reflections" (ie, related model data) into their
+      # own objects
+      serializer._reflections.each do |reflection_key, reflection|
+        next unless object.key?(reflection_key) && object[reflection_key].present?
+        results.deep_merge!(flatten(object[reflection_key], reflection.options[:serializer], reflection.name))
+      end
+    end
+    return results
+  end
+
+  # The ScriptCrowdinSerializer used in the sync in organizes all content by
+  # script; each script is a single file, containing all the individual
+  # lessons, activities, etc, for the script. This works great for organizing
+  # content in a way that can easily be worked on by translators, but for
+  # rendering translations in model and view logic we're going to need to do
+  # something different.
+  #
+  # Specifically, we want to flatten the complex files generated by the
+  # sync in. Ultimately, we want one file per content type; one for all
+  # lessons, one for all activities, one for all resources, etc.
+  def self.sync_out_reorganize
+    Languages.get_locale.each do |prop|
+      locale = prop[:locale_s]
+      next if locale == 'en-US'
+      next unless locale == 'fr-FR'
+      locale_dir = File.join('i18n/locales', locale, 'curriculum_content')
+      next unless File.directory?(locale_dir)
+
+      # First we gather together all our script objects into a single hash
+      script_objects = {}
+      Dir.glob(File.join(locale_dir, '**/*.json')).each do |file|
+        data = JSON.parse(File.read(file))
+        name = File.basename(file)
+        script_objects[name] = data if data.present?
+      end
+
+      # Then we recurisvely flatten all of our hashes of objects, and write
+      # each resulting collection of strings out to a rails i18n config file.
+      result = flatten(script_objects, ScriptCrowdinSerializer, :scripts)
+      result.each do |type, strings|
+        dest = File.join(Rails.root, 'config/locales', "#{type}.#{locale}.json")
+        data = {locale => {data: {type => strings}}}
+        # TODO: normalize keys for lessons and resources
+        # route_params = Rails.application.routes.recognize_path(crowdin_key)
+        # {:controller=>"lessons", :action=>"show", :script_id=>"coursea-2021", :position=>"1"}
```

Comment:
Is this all one comment or two? If you don't plan on fixing this TODO soon, could you leave an example of what a normalized key would look like so future developers know what you had in mind.

Reply:
oh, some of those are just stray notes.

I'm planning on resolving all of these TODOs myself, this very sprint